domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt

```diff
@@ -123,6 +132,33 @@ constructor(
 
     private lateinit var imageCaptureUseCase: ImageCapture
 
+    /**
+     * Applies a CaptureCallback to the provided image capture builder
+     */
+    @OptIn(ExperimentalCamera2Interop::class)
+    private fun onCaptureCompletedCallback(imageCaptureBuilder: ImageCapture.Builder) {
+        val captureCallback = object : CameraCaptureSession.CaptureCallback() {
+            override fun onCaptureCompleted(
+                session: CameraCaptureSession,
+                request: CaptureRequest,
+                result: TotalCaptureResult
+            ) {
+                super.onCaptureCompleted(session, request, result)
+                try {
+                    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
+                        Trace.endAsyncSection(FIRST_FRAME_TRACE, FIRST_FRAME_COOKIE)
```

Comment:
Having cross-module begin/end could be problematic if the code changes in one spot or the other. I would recommend plumbing up a callback to the spot where the trace is started. So instead of calling `Trace.endAsyncSection` here, you would just invoke the callback. It could be named something like `firstFrameComplete()` and can be passed into the `CameraUseCase.runCamera()` function.

That way you can call `begin/endSection()` at the same spot, and only need to define the trace tag in a single spot.

Reply:
I ended up creating a singleton TraceManager class to cleanly facilitate tracing anywhere in the app. Just inject it where you need it.
I think for future traces, it may become messy to repeatedly plumb up callbacks throughout the app, so I think this may save some future legibility